ouroboros-consensus/src/Ouroboros/Storage/ChainDB/Model.hs

```diff
@@ -184,8 +191,16 @@ addBlock cfg blk m
     blocks' :: Map (HeaderHash blk) blk
     blocks' = Map.insert (Block.blockHash blk) blk (blocks m)
 
-    candidates :: [(Chain blk, ExtLedgerState blk)]
-    candidates = mapMaybe (validate cfg (initLedger m)) $ chains blocks'
+    invalidBlocks' :: Set (Point blk)
+    candidates     :: [(Chain blk, ExtLedgerState blk)]
+    (invalidBlocks', candidates) =
+      foldMap (classify . validate cfg (initLedger m)) $ chains blocks'
```

Comment:
Since ledger states are involved here, does the memory use rely on the candidates being calculated incrementally? If so, is this still incremental? Returning a pair of a set and a list seems suspicious as it's very hard to do that such that the result can be consumed incrementally.